### PR TITLE
Initial commit of the rewritten registrar

### PR DESCRIPTION
**Pull request by JonathanO**

### CODE REVIEW

**Review by mihkelparna1 — APPROVED** [review]
Looks good just from a basic understanding of what's being done, don't have enough experience to comment from 'what's the best way of doing things' perspective though.

File: config.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+func LoadConfig(path string) (*Config, error) {
+	hclBytes, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, errs.New("unable to load configuration: %v", err)
+	}
+	return ParseConfig(string(hclBytes))
+}
+
+func ParseConfig(hclConfig string) (*Config, error) {
+	c := new(Config)
```
Comment (robinp-tw) [inline config.go]: nit: why not `c := Config{}`, and return as pointer later?

File: controllers/base_controller.go [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+						reconciled++
+						log.V(1).Info("Triggering reconciliation for resource", "name", namespacedName)
+						out <- event.GenericEvent{Meta: &v1.ObjectMeta{
+							Name:      namespacedName.Name,
+							Namespace: namespacedName.Namespace,
+						}}
+					}
+				}
+			}
+			log.Info("Synced spire entries", "took", time.Since(start), "found", len(entries), "queued", reconciled)
```
Comment (robinp-tw) [inline controllers/base_controller.go]: key name `queued` vs value `reconciled` ? Maybe.

File: controllers/node_controller.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+func (r *NodeReconciler) getAllEntries(ctx context.Context) ([]*common.RegistrationEntry, error) {
+	entries, err := r.SpireClient.ListByParentID(ctx, &registration.ParentID{Id: r.RootId})
+	if err != nil {
+		return nil, err
+	}
+	return entries.Entries, nil
+}
+
+func (r *NodeReconciler) getObject() ObjectWithMetadata {
+	return &corev1.Node{}
```
Comment (robinp-tw) [inline controllers/node_controller.go:68]: Comment if this is temporary or always fine to return empty Node (and maybe why)

**Review by robinp-tw — COMMENTED** [review]
Random nits.

File: controllers/node_controller.go (hunk 3 above)
Comment (JonathanO) [reply controllers/node_controller.go:68]: Oh. This. As far as I can tell this is how people do "generics" in Go, or at least how the k8s client library seems to implement them. The base controlled expects to call this and get an appropriate object type for the k8s client to populate.

**Pull request merged**